### PR TITLE
fix: incorrect pointer value comparison

### PR DESCRIPTION
**Pull request by ltzmaxwell**
This is a fix to #1569 . Thanks to @thehowl  for identifying this issue and providing a thorough and insightful analysis.

Here is an further analysis based on #1569 :

```go
package main

func main() {
	{
		b := []byte("ABCDEFGHIJKL")
		a := b
		println(&a[0] == &a[0])
	}
}
```
this comparison would be false too.

The root cause for this is the way pointer values is obtained and compared,  e.g. 
```go
package main

func main() {
	c := []byte{'A'}
	a := c
	println(&c[0])
	println(&c[0] == &a[0])
}
```

in this code snippet, the address of the c[0], (&c[0]) is obtained from this:
```go
    ev := fillValueTV(store, &av.List[ii]) // by reference
```
that is a reference to the element of the underlying list of arrayValue,

in case like this:
```go
package main

func main() {
	{
		b := []byte("ABCDEFGHIJKL")
		a := b
		println(&a[0] == &a[0])
	}
}
```
the address is obtained by
```go
	bv := &TypedValue{ // heap alloc, so need to compare value rather than pointer in isEql(), line 482
		T: DataByteType,
		V: DataByteValue{
			Base:     av,
			Index:    ii,
			ElemType: et,
		},
	}
```
that is a new allocated *TV, which implies the value of it would not be same within the first and second & operation.
So we should actually compare the concrete value rather than the pointers for this case.

### CODE REVIEW

File: gnovm/pkg/gnolang/op_expressions.go [hunk 1]
```diff
@@ -194,8 +194,12 @@ func (m *Machine) doOpRef() {
 			nv.Value = rv2
 		}
 	}
+	elt := xv.TV.T
```
Comment (deelawn) [inline gnovm/pkg/gnolang/op_expressions.go:197]: Are the changes in this file necessary? The `TV.T` doesn't appear to be consumed later during the equality check. The tests that have been added still pass with these changes excluded.

**Review by deelawn — CHANGES_REQUESTED** [review]
Looks pretty good.  I think we need to guard against invalid type comparisons.

**Comment by piux2** [comment]
I think we should convert []byte("ABCD") to  []byte{65, 66, 67, 68} in  preprocess.

**Comment by piux2** [comment]
I think we should try to find a way to convert []byte("ABCDEFG") to []byte{65, 66, 67, 68} in preprocess.

**Comment by zivkovicmilos** [comment]
@ltzmaxwell 
Can you please check @deelawn's comments on this PR? 🙏

File: gnovm/pkg/gnolang/values.go [hunk 2]
```diff
@@ -361,14 +361,15 @@ func (av *ArrayValue) GetPointerAtIndexInt2(store Store, ii int, et Type) Pointe
 			Index: ii,
 		}
 	}
-	bv := &TypedValue{ // heap alloc
+	bv := &TypedValue{ // heap alloc, so need to compare value rather than pointer in isEql(), line 482
```
[the remark this answers is not in the document]
Comment (ltzmaxwell) [reply gnovm/pkg/gnolang/values.go:364]: that make sense.

**Comment by ltzmaxwell** [comment]
> @ltzmaxwell Can you please check @deelawn's comments on this PR? 🙏

Updated 🙏 .

**Review by petar-dambovaliev — APPROVED** [review]

**Comment by deelawn** [comment]
> > @ltzmaxwell Can you please check @deelawn's comments on this PR? 🙏
> 
> Updated 🙏 .

I don't think that fixed it -- see this comment https://github.com/gnolang/gno/pull/1601#discussion_r1505134477

File: gnovm/pkg/gnolang/op_binary.go [hunk 3]
```diff
@@ -475,7 +475,13 @@ func isEql(store Store, lv, rv *TypedValue) bool {
 				rfv.GetClosure(store)
 		}
 	case PointerKind:
-		// TODO: assumes runtime instance normalization.
+		if lv.V != nil && rv.V != nil {
+			lpv := lv.V.(PointerValue)
+			rpv := rv.V.(PointerValue)
```
[the remark this answers is not in the document]
Comment (ltzmaxwell) [reply gnovm/pkg/gnolang/op_binary.go:480]: Thank you @deelawn for the review. 🙏 . This seems to be another problem. If the assertion failed the comparison should be invalid (invalid operation: mismatched type xxx and xxx). or, even rv.V is pointer value they can be incomparable either.

think:

package main

func main() {
intPtr := new(int8)
*intPtr = 5

i := 0
println(intPtr == &i)
}

I believe this is something fixed in https://github.com/gnolang/gno/pull/1426.

see here:
https://github.com/gnolang/gno/blob/5b83cdbfdbfd6e95e41b126b46ac5d62f5fe0088/gnovm/tests/files/types/0a02.gno

https://github.com/gnolang/gno/blob/3d148f6d3df828e3621c6670f743e312e73001a6/gnovm/tests/files/types/0a03.gno

WDYT?

**Review by deelawn — APPROVED** [review]

**Pull request merged**